### DISCUSSION TITLE
banktransfer: ability to enable/disable (country-specific) QR codes

### DISCUSSION THREAD

**Discussion by stapelberg** [pretix/pretix#2723] [Ideas]
The banktransfer plugin currently displays BezahlCode and EPC-QR codes in the pending payment view.

But, in Switzerland, my e-banking app can scan neither of these codes.

Instead, Switzerland uses the QR-bill standard, see for example my implementation for more details: https://github.com/stapelberg/qrbill

I have a patch which modifies `src/pretix/plugins/banktransfer/templates/pretixplugins/banktransfer/pending.html` to generate a QR-bill code instead. This is easy enough to do if you don’t mind diverging from the standard a tiny bit (can just replace the QR code textual contents), and a tiny bit more involved operationally (setting up and integrating my `qrbill` program/service) if you want to adhere to the standard exactly (e.g. swiss cross logo overlay and exact QR parameters specified by the standard).

I was wondering how we could best upstream this patch. Ideally, the banktransfer plugin would offer a few checkboxes in its settings page that influences which codes should be shown. Does that make sense? Would you accept a PR?

**Comment by raphaelm** [comment]
Hmm, it's not an easy topic to tackle and we need to take this in two steps.

## Hiding the existing codes

Generally, we try to add as little new checkboxes as possible and do as much automatic good as possible, because pretix already has more checkboxes than most people can reason about properly, even I forget sometimes what checkboxes we have.

Currently, we show two types of QR codes for SEPA accounts:

- BezahlCode, an older German standard that does not seem to be used a lot any more. We could maybe just try dropping it and see if people complain?
- EPC-QR, also known in Germany as "GiroCode", this is an official standard by the [European Payments Council](https://www.europeanpaymentscouncil.eu/sites/default/files/KB/files/EPC069-12%20v2.1%20Quick%20Response%20Code%20-%20Guidelines%20to%20Enable%20the%20Data%20Capture%20for%20the%20Initiation%20of%20a%20SCT.pdf) that seems to be in active use.

Note that you very likely selected the following option in your settings:

![2022-07-20T18:32:53+02:00](https://user-images.githubusercontent.com/64280/180036402-6a18c49f-e2dd-4d5a-ae28-2592bffa0099.png)

This is not entirely wrong, since Switzerland is technically a SEPA member state. However, if EPC QR is an official SEPA standard, I don't see why we should stop showing it just because your one Swiss banking app can't scan it, other Swiss apps might and apps from other SEPA countries very likely can.

If you really want to turn it off, with the "other bank account" option you can tell pretix not to apply assumptions about SEPA standards and the barcodes will disappear. I don't think you would lose any other *relevant* features unless you use our capabilities to export SEPA XML files for *refunds*, but then again, what are the chances of Swiss banks even being able to read these files.

## Adding QR-bill

Now adding QR-bill is an entirely different topic that we've asked about before and that we'd like to tackle at some point. We have not yet started it because it seems complex if you want to do it properly (and we like doing things properly):

- Technically, the QR-bill standards wants us not only to provide a QR code, but they want to control an entire 1/3 page of the invoice
  - which assumes there always is an invoice, which is not always a valid assumption in pretix
  - which assumes that the invoice is created *after* the payment method is chosen, which is not always a valid assumption in pretix
  - which assumes a certain layout of the invoice, which conflicts with e.g. footer lines
  - which includes a mandated choice of fonts, only one of which is a free font, which is annoying as we need to bundle it with pretix
- They mandate a Swiss cross in the middle of their QR code, I have trouble taking this seriously as a standard, and also it means we can't trivially use the same QR encoding library as for all of our other stuff

I'm sure it's totally fine to ignore all of this for a small event, but we have a few "serious business" customers in Switzerland who'd probably be worried about an implementation that does not conform to this standard, so I'd be cautious merging it into a pretix core module. (Also, for a pretix core module that we'll continue to maintain in the future we would not want to depend on a microservice or go binary for this and would need a python implementation, but that doesn't seem too hard to do if we figure the rest out)

**Reply by stapelberg** [reply]
Thanks for taking a look!

> This is not entirely wrong, since Switzerland is technically a SEPA member state. However, if EPC QR is an official SEPA standard, I don't see why we should stop showing it just because your one Swiss banking app can't scan it, other Swiss apps might and apps from other SEPA countries very likely can.

I tested it with the ZKB app and with the PostFinance app. Aside from these two, the only other big players are UBS and Credit Suisse. If you want, I can ask around and have folks test with those apps, too, but I bet they don’t accept EPC-QR either.

> If you really want to turn it off, with the "other bank account" option you can tell pretix not to apply assumptions about SEPA standards and the barcodes will disappear. I don't think you would lose any other relevant features unless you use our capabilities to export SEPA XML files for refunds, but then again, what are the chances of Swiss banks even being able to read these files.

Good to know! Yeah, I don’t think Swiss banks accept SEPA XML files.

> Technically, the QR-bill standards wants us not only to provide a QR code, but they want to control an entire 1/3 page of the invoice

I think that bit of the standard is — in practice — only relevant if you plan to send paper invoices to folks who then physically take it to their bank and want to have it scanned. And even then, my experience is that a non-standard invoice might raise an eyebrow and confuse the employee, but their QR code scanner doesn’t care for anything but the code of course.

> They mandate a Swiss cross in the middle of their QR code, I have trouble taking this seriously as a standard, […]

You can of course find it ridiculous, but the recognizability of QR bill codes is a feature I find quite nice. When I saw the codes on the pretix payment page, I immediately assumed they wouldn’t work (and they didn’t), and when I see a code with the swiss cross I assume it works (and they always did, in my experience, so far).

> […], and also it means we can't trivially use the same QR encoding library as for all of our other stuff

Why? In my qrbill implementation I overlay the swiss cross afterwards: https://github.com/stapelberg/qrbill/blob/05d1f60c14c6b0a6fbb84bb025036757c671feb0/qrcodegenerator.go#L81

Anyway, I’ll leave it up to you how you want to proceed. I have already spent many days on my qrbill implementation, so I’ll just use that with a custom pretix patch instead of re-implementing it all in Python (which I understand that you’d prefer for pretix obviously). Let me know if you’d like me to test your codes at some point.

**Reply by raphaelm** [reply]
> I think that bit of the standard is — in practice — only relevant if you plan to send paper invoices to folks who then physically take it to their bank and want to have it scanned. And even then, my experience is that a non-standard invoice might raise an eyebrow and confuse the employee, but their QR code scanner doesn’t care for anything but the code of course.

Thank you, that's a good perspective. I'll talk to 1-2 Swiss customers who requested QR-bill before, maybe we can really get away with a simple solution like this.

> I think that bit of the standard is — in practice — only relevant if you plan to send paper invoices to folks who then physically take it to their bank and want to have it scanned. And even then, my experience is that a non-standard invoice might raise an eyebrow and confuse the employee, but their QR code scanner doesn’t care for anything but the code of course.

Yeah, also a good point.

> Anyway, I’ll leave it up to you how you want to proceed. I have already spent many days on my qrbill implementation, so I’ll just use that with a custom pretix patch instead of re-implementing it all in Python (which I understand that you’d prefer for pretix obviously).

Seems like someone already did the nasty work for Python:
https://pypi.org/project/qrbill/

**Reply by raphaelm** [reply]
Our customer who requested this before agreed that just showing the barcode is already helpful at least for a B2C context.

I've created the most simple implementation here: https://github.com/pretix/pretix/pull/2767

I'd be delighted if you want to take it for a spin and test if it works for you, @stapelberg  :)

**Reply by stapelberg** [reply]
Thanks for the update! I’m traveling right now but will take a look at this over the weekend when I’ll be back.

**Reply by raphaelm** [reply]
Yeah, sure, there's no hurry on this, I'll be OOO from this weekend until mid September anyways, just wanted to cross this off on my to-do list beforehand which I've done now by creating the PR :D
